packages/core/src/Parser.js

```diff
@@ -5,7 +5,7 @@ const Promise = require('bluebird');
 const slugify = require('@sindresorhus/slugify');
 const componentParser = require('./parsers/componentParser');
 const componentPreprocessor = require('./preprocessors/componentPreprocessor');
-const logger = require('../../../util/logger');
+const logger = require('../../../src/util/logger');
```

Comment:
Hmm I was thinking of dealing with this in the pr that would move all the non-cli things inside `core`, since most files using it would be moved there. For now just a simple addition of `src` to the import path to avoid breaking things

Reply:
I realised I didn't add `ci:all` ( and the discussed `npm run build:components` ) to netlify during #1255, likely why the netlify build is failing
Could I trouble you to add the necessary permissions? 😅

Reply:
Using a published `@markbind/core` would break.

Even if we *move all the non-cli things inside `core`*, I think logging to a file should remain in `markbind-cli`.

Reply:
I don't have access to Netlify. @damithc?

Reply:
Ahh good point, just realised how this works as well, I'll add it as the 3rd commit after `Change import paths to use @markbind/core` then!

> Looks like we can do winston.configure(...) in packages/core/src/logger.js and then
winston.add(new DailyRotateFile(...)) in src/util/logger.js.

Reply:
> Looks like we can do winston.configure(...) in packages/core/src/logger.js and then
winston.add(new DailyRotateFile(...)) in src/util/logger.js.

Actually, not sure if this is a good way to structure the logging. This is reliant on `logger.info` not being used in `core` (since the default logger instance is being used), once we move `Site.js` this wouldn't be true anymore. We may also add `logger.info` to `Parser.js` and such.

I wonder if we should use dependency injection for logger in the core package instead, this would also make sense from the perspective of a future `markbind-ui`.

e.g.

```
In core package,

new Site(..., logger)


In markbind-cli,

pass logger instance that logs to both file and console

In markbind-ui,

pass logger instance that does not have console in its transports (only logging to the file)
```

The downside being needing to pass `logger` to `Site/Page/Parser..../down the composition chain`.

Not sure if there's a better solution, or if I misunderstood what you meant by:

> Looks like we can do winston.configure(...) in packages/core/src/logger.js and then
winston.add(new DailyRotateFile(...)) in src/util/logger.js.

Reply:
We could also do:

```
In core package logger.js,

module.exports = {
  ... logger function stubs ...
}

In markbind-cli/ui,

require('@markbind/core').logger, then patch the stubs
```

This avoids having to pass `logger` down through `Site/Page/...`, but seems a little unconventional

Reply:
> This is reliant on `logger.info` not being used in `core` (since the default logger instance is being used), once we move `Site.js` this wouldn't be true anymore.

1. What's the link?
<br>

> We may also add `logger.info` to `Parser.js` and such.

2. Why would you do that?
<br>

> I wonder if we should use dependency injection for logger in the core package instead, this would also make sense from the perspective of a future `markbind-ui`.

3. What's `markbind-ui`?
(Btw, MarkBind doesn't use a dependency injection framework, so I think you mean *accept a constructor argument*.)
<br>

> `require('@markbind/core').logger, then patch the stubs`

4. What's there to patch?

---

> Not sure if there's a better solution, or if I misunderstood what you meant by:

I mean:

```js
// packages/core/src/logger.js

// Pretty much the existing src/util/logger.js without DailyRotateFile
```

```js
// src/util/logger.js

const logger = require('@markbind/core/src/logger');
const winston = require('winston');
const DailyRotateFile = require('winston-daily-rotate-file');

winston.add(new DailyRotateFile({
  // ...
}));

module.exports = logger;
```

Reply:
1. 

> What's the link?

My bad some more context: the issue is in the logging levels, currently this is what we have:

```
transports: [
    new (winston.transports.Console)({
      colorize: true,
      handleExceptions: true,
      humanReadableUnhandledException: true,
      level: 'info',
      showLevel: true,
    }),
    new DailyRotateFile({
      datePattern: 'YYYY-MM-DD',
      dirname: '_markbind/logs',
      filename: 'markbind-%DATE%.log',
      handleExceptions: true,
      humanReadableUnhandledException: true,
      level: 'debug',
      maxFiles: 5,
      showLevel: true,
    }),
  ],
```

Hence `http/verbose/debug` level messages get output into the file but not console, which is a behaviour we want to preserve (we shouldn't spew debug level things out into the console)

From a perspective of unit tests, we want: all `logger.xx` calls to show somewhere at least once. 
`level: info` blocks `debug` level messages from showing, so we have to change it to `level: debug`, however:

From a perspective of the author, we want: `http/verbose/debug` level messages to get output into the file but not the console. (the user shouldn't be spammed with `debug` level messages)

> so we have to change it to `level: debug`

The above makes this impossible


2. Ignore this 😅

3.

>What's markbind-ui?

Just a fleeting comment I saw in an unrelated issue, I suppose it would be the `ui` equivalent of `markbind-cli` providing the same functionality.

> Btw, MarkBind doesn't use a dependency injection framework, so I think you mean accept a constructor argument.

Yes, we are still performing it though. A framework isn't required, although often present

4.

> What's there to patch?

Something like that:

```
In Parser.js,

logger = require('src/utils/logger') -- stub / simple console implementation

In markbind-cli index.js,

coreLogger = require('@markbind/core/src/utils/logger');
coreLogger.info = () => winston.info(...)
coreLogger.error = ...
coreLogger.debug = ...
```

Reply:
realised I mistakenly closed #1207 as well: `logger` dosen't work in unit tests where we mock `fs`

Would it be better to just deal with this in a separate pr, since it's rather problematic and we aren't publishing `@markbind/core`?

Reply:
> From a perspective of the author, we want: `http/verbose/debug` level messages to get output into the file but not the console. (the user shouldn't be spammed with `debug` level messages)

This is based on usage of the CLI.
In the context of `@markbind/core`, we do not(?) assume that there is a writeable local filesystem.

> From a perspective of unit tests, we want: all `logger.xx` calls to show somewhere at least once. 
`level: info` blocks `debug` level messages from showing, so we have to change it to `level: debug`

> The above makes this impossible

>> What's there to patch?
>
> Something like that:

Actually, we can just do `winston.configure` again (i.e. replace) in `src/util/logger.js`.

> > Btw, MarkBind doesn't use a dependency injection framework, so I think you mean accept a constructor argument.
> 
> Yes, we are still performing it though. A framework isn't required, although often present

Erm no, by definition, passing the instance explicitly is not performing dependency injection.

Reply:
> This is based on usage of the CLI.
> In the context of @markbind/core, we do not(?) assume that there is a writeable local filesystem.

Yup, there isn't a opposite of "`winston.add(...)`" though, which is the issue in trying to achieve both perspectives' behaviours.



> Actually, we can just do winston.configure again (i.e. replace) in src/util/logger.js.

This easily subjects it the configuration to import order when using from author side:
ie., if `@markbind/core/src/utils/logger` (in fact, likely since it is deeper within the `require` chain) loads later, the earlier `markbind-cli` `winston.configure` is overwritten, which isn't what we want =\

Winston specifically provides `winston.createLogger` likely to get around cases such as this.
I don't think it intended for a one-size-fits-all logger; The `createLogger` seems to be mentioned as the first option in it's docs as well https://www.npmjs.com/package/winston

> Erm no, by definition, passing the instance explicitly is not performing dependency injection.

Pretty sure it qualifies https://nus-cs2103-ay1920s2.github.io/website/se-book-adapted/chapters/testing.html#dependency-injection o_o, we're injecting `logger` as a service from `markbind-cli` _or_ `markbind-ui` to modify the behaviour

https://en.wikipedia.org/wiki/Dependency_injection#:~:text=In%20software%20engineering%2C%20dependency%20injection,object%20is%20called%20a%20service.

It's just that in many back-end frameworks (laravel, .net, ...) you'd also have a "service container" that holds all the _many_ things behind the scenes to inject, this is not necessitated by the definition:

> The code that passes the service to the client can be many kinds of things

https://en.wikipedia.org/wiki/Dependency_injection#:~:text=In%20software%20engineering%2C%20dependency%20injection,object%20is%20called%20a%20service.

Reply:
I really think we should deal with logging in an entirely separate PR though 😅. @nbriannl and I grappled with it a while back in #1207 (which is still open) as well, it's a deceptively simple issue that we shouldn't touch here

Reply:
> > Actually, we can just do winston.configure again (i.e. replace) in src/util/logger.js.
> 
> This easily subjects it the configuration to import order when using from author side:
> ie., if `@markbind/core/src/utils/logger` (in fact, likely since it is deeper within the `require` chain) loads later, the earlier `markbind-cli` `winston.configure` is overwritten, which isn't what we want =\

Not if `markbind-cli`'s `src/util/logger.js` does `require('@markbind/core/src/utils/logger')` before calling `winston.configure`.

> I really think we should deal with logging in an entirely separate PR though 😅. @nbriannl and I grappled with it a while back in #1207 (which is still open) as well, it's a deceptively simple issue that we shouldn't touch here

Do you mean #1060? The solution suggested above should keep it independent of `DailyRotateFile` right?

---

> > Erm no, by definition, passing the instance explicitly is not performing dependency injection.
> 
> Pretty sure it qualifies https://nus-cs2103-ay1920s2.github.io/website/se-book-adapted/chapters/testing.html#dependency-injection o_o, we're injecting `logger` as a service from `markbind-cli` _or_ `markbind-ui` to modify the behaviour
> 
> https://en.wikipedia.org/wiki/Dependency_injection#:~:text=In%20software%20engineering%2C%20dependency%20injection,object%20is%20called%20a%20service.
> 
> It's just that in many back-end frameworks (laravel, .net, ...) you'd also have a "service container" that holds all the _many_ things behind the scenes to inject, this is not necessitated by the definition:

Oops, I stand corrected. Thank you for pointing that out.
I had the misconception that the "injector" role had to be separate from the requestor.

Reply:
> Not if markbind-cli's src/util/logger.js does require('@markbind/core/src/utils/logger') before calling winston.configure.

Ahh that makes sense! I'll give it a try

> Do you mean #1060? The solution suggested above should keep it independent of DailyRotateFile right?

yup

Reply:
updated @acjh, works well 👍, we have proper output in unit tests again

@damithc could you add the necessary netlify permissions to `@MarkBind/admins`?

Reply:
> @damithc could you add the necessary netlify permissions to `@MarkBind/admins`?

Not sure how to do that. But I've invited you to the MarkBind's Netlify account.